File: examples/python/dtos.py
from dataclasses import dataclass
from datetime import datetime, date

# Sempre deixe o dict gerado com nomes camelCase!

@dataclass
class User:
    name: str
    birth_date: date

    def to_dict(self) -> dict:
        birth_str = self.birth_date.strftime('%d/%m/%Y') if isinstance(self.birth_date, (date, datetime)) else str(self.birth_date)
        return {
            'name': self.name,
            'birthDate': birth_str
        }

    @classmethod
    def from_dict(cls, data: dict) -> 'User':
        birth_date_val = data.get('birthDate')
        if isinstance(birth_date_val, str):
            parsed_date = datetime.strptime(birth_date_val, '%d/%m/%Y').date()
        elif isinstance(birth_date_val, datetime):
            parsed_date = birth_date_val.date()
        elif isinstance(birth_date_val, date):
            parsed_date = birth_date_val
        else:
            parsed_date = date.today()
            
        return cls(
            name=data.get('name', ''),
            birth_date=parsed_date
        )


@dataclass
class UserAstrology(User):
    week_day: str = ''
    sign: str = ''

    def to_dict(self) -> dict:
        data = super().to_dict()
        data.update({
            'weekDay': self.week_day,
            'sign': self.sign
        })
        return data

    @classmethod
    def from_dict(cls, data: dict) -> 'UserAstrology':
        user = super().from_dict(data)
        return cls(
            name=user.name,
            birth_date=user.birth_date,
            week_day=data.get('weekDay', ''),
            sign=data.get('sign', '')
        )


@dataclass
class Talk:
    author: str
    text: str
    reply: bool = False

    def to_dict(self) -> dict:
        return {
            'author': self.author,
            'text': self.text,
            'reply': self.reply
        }

    @classmethod
    def from_dict(cls, data: dict) -> 'Talk':
        return cls(
            author=data.get('author', ''),
            text=data.get('text', ''),
            reply=data.get('reply', False)
        )

File: examples/dart/lib/dtos.dart
class User {
  final String name;
  final DateTime birthDate;

  User({required this.name, required this.birthDate});

  Map<String, dynamic> toJson() {
    final day = birthDate.day.toString().padLeft(2, '0');
    final month = birthDate.month.toString().padLeft(2, '0');
    final year = birthDate.year;

    return {'name': name, 'birthDate': '$day/$month/$year'};
  }

  factory User.fromJson(Map<String, dynamic> json) {
    final birthDateVal = json['birthDate'];
    DateTime parsedDate;

    if (birthDateVal is String && birthDateVal.isNotEmpty) {
      final parts = birthDateVal.split('/');
      if (parts.length == 3) {
        final day = int.parse(parts[0]);
        final month = int.parse(parts[1]);
        final year = int.parse(parts[2]);
        parsedDate = DateTime(year, month, day);
      } else {
        parsedDate = DateTime.parse(birthDateVal);
      }
    } else {
      parsedDate = DateTime.now();
    }

    return User(name: json['name'] as String? ?? '', birthDate: parsedDate);
  }

  @override
  String toString() {
    return 'User(name: $name, birthDate: ${toJson()['birthDate']})';
  }
}

class UserAstrology extends User {
  final String weekDay;
  final String sign;

  UserAstrology({required super.name, required super.birthDate, this.weekDay = '', this.sign = ''});

  @override
  Map<String, dynamic> toJson() {
    final data = super.toJson();
    data['weekDay'] = weekDay;
    data['sign'] = sign;
    return data;
  }

  factory UserAstrology.fromJson(Map<String, dynamic> json) {
    final user = User.fromJson(json);
    return UserAstrology(
      name: user.name,
      birthDate: user.birthDate,
      weekDay: json['weekDay'] as String? ?? '',
      sign: json['sign'] as String? ?? '',
    );
  }

  @override
  String toString() {
    return 'UserAstrology(name: $name, birthDate: ${toJson()['birthDate']}, weekDay: $weekDay, sign: $sign)';
  }
}

class Talk {
  final String author;
  final String text;
  final bool reply;

  Talk({required this.author, required this.text, this.reply = false});

  Map<String, dynamic> toJson() {
    return {'author': author, 'text': text, 'reply': reply};
  }

  factory Talk.fromJson(Map<String, dynamic> json) {
    return Talk(
      author: json['author'] as String? ?? '',
      text: json['text'] as String? ?? '',
      reply: json['reply'] as bool? ?? false,
    );
  }

  @override
  String toString() {
    return 'Talk(author: $author, text: $text, reply: $reply)';
  }
}
